x/auth/tx/builder.go

```diff
@@ -109,25 +104,23 @@ func (w *wrapper) GetSigners() []sdk.AccAddress {
 }
 
 func (w *wrapper) GetPubKeys() []crypto.PubKey {
-	if w.pubKeys == nil {
-		signerInfos := w.tx.AuthInfo.SignerInfos
-		pubKeys := make([]crypto.PubKey, len(signerInfos))
+	signerInfos := w.tx.AuthInfo.SignerInfos
+	pks := make([]crypto.PubKey, len(signerInfos))
 
-		for i, si := range signerInfos {
-			var err error
-			pk := si.PublicKey
-			if pk != nil {
-				pubKeys[i], err = w.pubkeyCodec.Decode(si.PublicKey)
-				if err != nil {
-					panic(err)
-				}
-			}
+	for i, si := range signerInfos {
+		// NOTE: it is okay to leave this nil if there is no PubKey in the SignerInfo.
+		// PubKey's can be left unset in SignerInfo.
+		if si.PublicKey == nil {
+			continue
 		}
 
-		w.pubKeys = pubKeys
+		pk, ok := si.PublicKey.GetCachedValue().(crypto.PubKey)
+		if ok {
+			pks[i] = pk
+		}
 	}
 
-	return w.pubKeys
+	return pks
```

Comment:
Some entries in `pks` will be "zero". Is it OK?

Reply:
I believe there's a comment already

> // NOTE: it is okay to leave this nil if there is no PubKey in the SignerInfo.
		// PubKey's can be left unset in SignerInfo.